In [80]:
import pandas as pd
import numpy as np
import requests

In [81]:
df = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]

In [82]:
df.head(12)

Postal Code           Borough                                 Neighborhood
0          M1A      Not assigned                                 Not assigned
1          M2A      Not assigned                                 Not assigned
2          M3A        North York                                    Parkwoods
3          M4A        North York                             Victoria Village
4          M5A  Downtown Toronto                    Regent Park, Harbourfront
5          M6A        North York             Lawrence Manor, Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7          M8A      Not assigned                                 Not assigned
8          M9A         Etobicoke      Islington Avenue, Humber Valley Village
9          M1B       Scarborough                               Malvern, Rouge
10         M2B      Not assigned                                 Not assigned
11         M3B        North York                                    Don Mills

In [83]:
indexNames = df[df['Borough'] == 'Not assigned'].index

df1 = df.drop (indexNames, inplace=True)

In [84]:
indexNamesNeigh = df[df['Neighborhood'] == 'Not assigned'].index
indexNamesNeigh


Int64Index([], dtype='int64')

In [85]:
df.shape

(103, 3)

I read the table in wikipedia using pandas and put it into a frame
i removed the rows where the values of borough are not assigned
i checked if any neighborhood values has a not assigned value and still available
i used df.shape in order to know that we have 103 rows left.

In [63]:
!conda install -c conda-forge geocoder

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geocoder


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ratelim-0.1.6              |             py_2           6 KB  conda-forge
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    geocoder-1.38.1            |             py_1          53 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geocoder:        1.38.1-py_1       conda-forge
    python_abi:    

In [86]:
dfll=pd.read_csv('http://cocl.us/Geospatial_data')

In [68]:
dfll.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [92]:
dfmerged = pd.merge(df, dfll, on="Postal Code")

In [93]:
dfmerged.shape

(103, 5)

In [95]:
dfmerged.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [104]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    branca-0.4.1               |             py_0          26 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         713 KB

The following NEW packages will be INSTALLED:

    altair:  4.1.0-py_1 conda-forge
    branca:  0.4.1-py_0 conda-forge
    folium:  0.5.0-py_0 conda-forge
    vincent: 0.4.4-py_1 conda-forge


branca-0.4.1         | 26 KB     | ##################################### | 1

In [107]:
dfTor = dfmerged[dfmerged["Borough"].str.contains('Toronto')]

In [106]:
dfTor.head()

Postal Code           Borough                                 Neighborhood  \
2          M5A  Downtown Toronto                    Regent Park, Harbourfront   
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9          M5B  Downtown Toronto                     Garden District, Ryerson   
15         M5C  Downtown Toronto                               St. James Town   
19         M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  
2   43.654260 -79.360636  
4   43.662301 -79.389494  
9   43.657162 -79.378937  
15  43.651494 -79.375418  
19  43.676357 -79.293031

In [108]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(dfTor['Borough'].unique()),
        dfTor.shape[0]
    )
)

The dataframe has 4 boroughs and 39 neighborhoods.


In [111]:
address = 'Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [114]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(dfTor['Latitude'], dfTor['Longitude'], dfTor['Borough'], dfTor['Neighborhood']):
    label = '{}, {}'.format(dfTor, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [123]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    LIMIT = 10
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [115]:
CLIENT_ID = 'X1KEMZVNHOHDSGVPA1PDVZ230XV4WMBCAF40MTPTLZU1IBQV' # your Foursquare ID
CLIENT_SECRET = 'MDS0CPLTITYU031STRCPRDYG33NVSO0JGONQMWDPNF2NXOTK' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: X1KEMZVNHOHDSGVPA1PDVZ230XV4WMBCAF40MTPTLZU1IBQV
CLIENT_SECRET:MDS0CPLTITYU031STRCPRDYG33NVSO0JGONQMWDPNF2NXOTK


In [125]:
# type your answer here

toronto_venues = getNearbyVenues(names=dfTor['Neighborhood'],
                                   latitudes=dfTor['Latitude'],
                                   longitudes=dfTor['Longitude']
                                  )
print("Done")


Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [127]:
print(toronto_venues.shape)
toronto_venues.head()

(348, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

In [128]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            10   
Brockton, Parkdale Village, Exhibition Place                           10   
Business reply mail Processing Centre, South Ce...                     10   
CN Tower, King and Spadina, Railway Lands, Harb...                     10   
Central Bay Street                                                     10   
Christie                                                               10   
Church and Wellesley                                                   10   
Commerce Court, Victoria Hotel                                         10   
Davisville                                                             10   
Davisville North                                                        8   
Dufferin, Dovercourt Village                                           10   
First Canadian Place, Underground city                                 10   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                               10   
Harbourfront East, Union Station, Toronto Islands                      10   
High Park, The Junction South                                          10   
India Bazaar, The Beaches West                                         10   
Kensington Market, Chinatown, Grange Park                              10   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               10   
Moore Park, Summerhill East                                             3   
North Toronto West, Lawrence Park                                      10   
Parkdale, Roncesvalles                                                 10   
Queen's Park, Ontario Provincial Government                            10   
Regent Park, Harbourfront                                              10   
Richmond, Adelaide, King                                               10   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     10   
St. James Town                                                         10   
St. James Town, Cabbagetown                                            10   
Stn A PO Boxes                                                         10   
Studio District                                                        10   
Summerhill West, Rathnelly, South Hill, Forest ...                     10   
The Annex, North Midtown, Yorkville                                    10   
The Beaches                                                             4   
The Danforth West, Riverdale                                           10   
Toronto Dominion Centre, Design Exchange                               10   
University of Toronto, Harbord                                         10   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             10   
Brockton, Parkdale Village, Exhibition Place                            10   
Business reply mail Processing Centre, South Ce...                      10   
CN Tower, King and Spadina, Railway Lands, Harb...                      10   
Central Bay Street                                                      10   
Christie                                                                10   
Church and Wellesley                                                    10   
Commerce Court, Victoria Hotel                                          10   
Davisville                                                              10   
Davisville North                    

In [129]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Lounge  Airport Terminal  \
0            0        0                   0               0                 0   
1            0        0                   0               0                 0   
2            0        0                   0               0                 0   
3            0        0                   0               0                 0   
4            0        0                   0               0                 0   

   American Restaurant  Arts & Crafts Store  Asian Restaurant  Auto Workshop  \
0                    0                    0                 0              0   
1                    0                    0                 0              0   
2                    0                    0                 0              0   
3                    0                    0                 0              0   
4                    0                    0                 0              0   

   BBQ Joint  Bakery  Bank  Bar  Beer Bar  Beer Store  Bookstore  Boutique  \
0          0       1     0    0         0           0          0         0   
1          0       0     0    0         0           0          0         0   
2          0       0     0    0         0           0          0         0   
3          0       0     0    0         0           0          0         0   
4          0       0     0    0         0           0          0         0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Burger Joint  Burrito Place  \
0               0        0                0             0              0   
1               0        0                0             0              0   
2               1        0                0             0              0   
3               0        0                0             0              0   
4               0        0                0             0              0   

   Bus Line  Café  Candy Store  Chinese Restaurant  Clothing Store  \
0         0     0            0                   0               0   
1         0     0            0                   0               0   
2         0     0            0                   0               0   
3         0     0            0                   0               0   
4         0     0            0                   0               0   

   Cocktail Bar  Coffee Shop  Comic Shop  Concert Hall  Cosmetics Shop  \
0             0            0           0             0               0   
1             0            1           0             0               0   
2             0            0           0             0               0   
3             0            0           0             0               0   
4             0            0           0             0               0   

   Creperie  Cuban Restaurant  Dance Studio  Department Store  Dessert Shop  \
0         0                 0             0                 0             0   
1         0                 0             0                 0             0   
2         0                 0             0                 0             0   
3         0                 0             0                 0             0   
4         0                 0             0                 0             0   

   Diner  Distribution Center  Dog Run  Donut Shop  \
0      0                    0        0           0   
1      0                    0        0           0   
2      0                    0        0           0   
3      0                    1        0           0   
4      0                    0        0           0   

   Eastern European Restaurant  Falafel Restaurant  Farmers Market  \
0                            0                   0               0   
1                            0                   0               0   
2                            0                   0               0   
3                            0                   0               0   
4                            0                   0               0   

   Fast Food Restauran

In [130]:
toronto_onehot.shape

(348, 120)

In [131]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  Airport  \
0                                         Berczy Park          0.0      0.0   
1        Brockton, Parkdale Village, Exhibition Place          0.0      0.0   
2   Business reply mail Processing Centre, South C...          0.0      0.0   
3   CN Tower, King and Spadina, Railway Lands, Har...          0.0      0.1   
4                                  Central Bay Street          0.0      0.0   
5                                            Christie          0.0      0.0   
6                                Church and Wellesley          0.0      0.0   
7                      Commerce Court, Victoria Hotel          0.0      0.0   
8                                          Davisville          0.0      0.0   
9                                    Davisville North          0.0      0.0   
10                       Dufferin, Dovercourt Village          0.0      0.0   
11             First Canadian Place, Underground city          0.0      0.0   
12    Forest Hill North & West, Forest Hill Road Park          0.0      0.0   
13                           Garden District, Ryerson          0.0      0.0   
14  Harbourfront East, Union Station, Toronto Islands          0.0      0.0   
15                      High Park, The Junction South          0.0      0.0   
16                     India Bazaar, The Beaches West          0.0      0.0   
17          Kensington Market, Chinatown, Grange Park          0.0      0.0   
18                                      Lawrence Park          0.0      0.0   
19                           Little Portugal, Trinity          0.0      0.0   
20                        Moore Park, Summerhill East          0.0      0.0   
21                  North Toronto West, Lawrence Park          0.1      0.0   
22                             Parkdale, Roncesvalles          0.0      0.0   
23        Queen's Park, Ontario Provincial Government          0.1      0.0   
24                          Regent Park, Harbourfront          0.0      0.0   
25                           Richmond, Adelaide, King          0.0      0.0   
26                                           Rosedale          0.0      0.0   
27                                           Roselawn          0.0      0.0   
28                                 Runnymede, Swansea          0.0      0.0   
29                                     St. James Town          0.0      0.0   
30                        St. James Town, Cabbagetown          0.0      0.0   
31                                     Stn A PO Boxes          0.0      0.0   
32                                    Studio District          0.0      0.0   
33  Summerhill West, Rathnelly, South Hill, Forest...          0.0      0.0   
34                The Annex, North Midtown, Yorkville          0.0      0.0   
35                                        The Beaches          0.0      0.0   
36                       The Danforth West, Riverdale          0.1      0.0   
37           Toronto Dominion Centre, Design Exchange          0.0      0.0   
38                     University of Toronto, Harbord          0.1      0.0   

    Airport Food Court  Airport Lounge  Airport Terminal  American Restaurant  \
0                  0.0             0.0               0.0                  0.0   
1                  0.0             0.0               0.0                  0.0   
2                  0.0             0.0               0.0                  0.0   
3                  0.1             0.2               0.1                  0.0   
4                  0.0             0.0               0.0                  0.0   
5                  0.0             0.0               0.0                  0.0   
6                  0.0             0.0               0.0                  0.0   
7                  0.0             0.0               0.0                  0.0   
8                  0.0             0.0               0.0                  0.0   
9                  0.0             0.0               0.0              

In [132]:
toronto_grouped.shape

(39, 120)

In [133]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                           venue  freq
0                   Liquor Store   0.1
1                       Fountain   0.1
2                           Park   0.1
3  Vegetarian / Vegan Restaurant   0.1
4                   Concert Hall   0.1


----Brockton, Parkdale Village, Exhibition Place----
                    venue  freq
0             Coffee Shop   0.2
1                     Gym   0.1
2  Furniture / Home Store   0.1
3              Restaurant   0.1
4                    Café   0.1


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0           Pizza Place   0.1
1         Auto Workshop   0.1
2         Burrito Place   0.1
3            Skate Park   0.1
4  Fast Food Restaurant   0.1


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0      Airport Lounge   0.2
1  Airport Food Court   0.1
2    Airport Terminal  

In [134]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [135]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place           Coffee Shop   
2  Business reply mail Processing Centre, South C...         Garden Center   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue   3rd Most Common Venue       4th Most Common Venue  \
0        Farmers Market            Concert Hall                Cocktail Bar   
1             Pet Store  Furniture / Home Store          Italian Restaurant   
2         Auto Workshop             Pizza Place               Burrito Place   
3           Coffee Shop                     Bar                     Airport   
4             Gastropub      Italian Restaurant  Modern European Restaurant   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                     Museum            Restaurant          Liquor Store   
1                        Gym            Restaurant                   Bar   
2                 Restaurant               Brewery            Skate Park   
3         Airport Food Court      Airport Terminal              Boutique   
4  Middle Eastern Restaurant       Bubble Tea Shop                   Spa   

  8th Most Common Venue          9th Most Common Venue 10th Most Common Venue  
0                  Park  Vegetarian / Vegan Restaurant               Fountain  
1                  Café                         Bakery                  Diner  
2        Farmers Market           Fast Food Restaurant             Comic Shop  
3                 Plane                Harbor / Marina     Falafel Restaurant  
4              Fountain                        Dog Run                 Garden

In [136]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 2, 0, 1, 1, 2, 0, 2, 1, 0], dtype=int32)

In [140]:
# add clustering labels
#neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = dfTor

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
2          M5A  Downtown Toronto                    Regent Park, Harbourfront   
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9          M5B  Downtown Toronto                     Garden District, Ryerson   
15         M5C  Downtown Toronto                               St. James Town   
19         M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
2   43.654260 -79.360636               0        Breakfast Spot   
4   43.662301 -79.389494               1           Yoga Studio   
9   43.657162 -79.378937               2                  Café   
15  43.651494 -79.375418               1           Coffee Shop   
19  43.676357 -79.293031               2                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2            Coffee Shop  Gym / Fitness Center   Distribution Center   
4       Sushi Restaurant    Italian Restaurant   Distribution Center   
9         Clothing Store              Tea Room                 Plaza   
15             Gastropub            Food Truck              Creperie   
19                   Pub     Health Food Store              Wine Bar   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2                  Restaurant         Historic Site                Bakery   
4          Mexican Restaurant              Creperie                  Park   
9               Burrito Place            Comic Shop           Music Venue   
15  Middle Eastern Restaurant            Restaurant   Japanese Restaurant   
19         Falafel Restaurant        Cosmetics Shop              Creperie   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
2                    Spa                  Park             Donut Shop  
4               Beer Bar           Coffee Shop    Arts & Crafts Store  
9        Thai Restaurant               Theater    Distribution Center  
15        Cosmetics Shop                   Gym               Fountain  
19      Cuban Restaurant          Dance Studio       Department Store

In [141]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters